### PR TITLE
Fixes #31790 - always delegate bootserver IP resolution to smart proxy

### PR DESCRIPTION
**Pull request by upadhyeammit**
In 60c03b6a2e92c432ee66f2b27da634954c1b646c (reverse) DNS search for TFTP server (bootServer) was delegated to the Smart Proxy if was supported, as signaled via the dhcp_filename_hostname capability. At the same time, in Foreman 2.0, the fallback support was deprecated. This drops the compatibility code.

### CODE REVIEW

**Comment by upadhyeammit** [comment]
@lzap I was checking original https://github.com/theforeman/foreman/pull/6465 and as per that we were suppose to remove deprecation block of code. As I see we are still having that code, I think we should update the version in deprecation warning?

**Comment by lzap** [comment]
Please do this:

```diff
diff --git a/app/models/concerns/orchestration/dhcp.rb b/app/models/concerns/orchestration/dhcp.rb
index dd77081ef..34a3e9bb4 100644
--- a/app/models/concerns/orchestration/dhcp.rb
+++ b/app/models/concerns/orchestration/dhcp.rb
@@ -92,14 +92,7 @@ module Orchestration::DHCP
       # we no longer convert the boot server to IPv4 - smart proxy modules are required to do so if they don't support hostname
       bs
     else
-      begin
-        Foreman::Deprecation.deprecation_warning('1.25', "DHCP proxy does not report dhcp_filename_* capability and reverse DNS search in Foreman will be removed.")
-        ip = NicIpResolver.new(:nic => self).to_ip_address(bs) if bs.present?
-        return ip unless ip.nil?
-        failure _("Unable to determine the host's boot server. The DHCP smart proxy failed to provide this information and this subnet is not provided with TFTP services.")
-      rescue => e
-        failure _("failed to detect boot server: %s") % e, e
-      end
+      failure _("Unable to determine boot server. The DHCP smart proxy doesn't provide dhcp_filename capability or failed to provide this information.")
     end
   end
 ```

We should now error out if boot server can't be detected. @ekohl ?

**Comment by ekohl** [comment]
Yes, 1.25 turned into 2.0. Since develop is now 2.5, that has long passed and it's safe to make it a hard error.

**Comment by upadhyeammit** [comment]
@lzap changes pushed, request to check. Thank you!

**Review by lzap — DISMISSED** [review]
The will do it, thanks.

File: app/models/concerns/orchestration/dhcp.rb [hunk 1]
```diff
@@ -92,14 +92,7 @@ def boot_server
...
     else
-      begin
-        Foreman::Deprecation.deprecation_warning('1.25', "DHCP proxy does not report dhcp_filename_* capability and reverse DNS search in Foreman will be removed.")
-        ip = NicIpResolver.new(:nic => self).to_ip_address(bs) if bs.present?
-        return ip unless ip.nil?
-        failure _("Unable to determine the host's boot server. The DHCP smart proxy failed to provide this information and this subnet is not provided with TFTP services.")
-      rescue => e
-        failure _("failed to detect boot server: %s") % e, e
-      end
+      failure _("Unable to determine boot server. The DHCP smart proxy doesn't provide dhcp_filename capability or failed to provide this information.")
```
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb]: I would consider moving this to just below `return unless tftp?` so the method becomes:
```ruby
return unless tftp?

unless subnet.dhcp.has_capability?(:DHCP, :dhcp_filename_hostname)
  failure ...
end

tftp.bootServer || URI.parse(subnet.tftp.url).host
```

For simplicity I left out the comments, but those can certainly help.

File: app/models/concerns/orchestration/dhcp.rb [hunk 2]
```diff
@@ -82,25 +82,15 @@ def boot_server
...
 
-    # first try to ask our IPv4 TFTP server for its boot server
-    bs = tftp.bootServer
-    # if that failed, trying to guess out tftp next server based on the smart proxy hostname
-    bs ||= URI.parse(subnet.tftp.url).host
+    unless subnet.dhcp.has_capability?(:DHCP, :dhcp_filename_hostname)
+      failure _("Unable to determine boot server. The DHCP smart proxy doesn't provide dhcp_filename capability or failed to provide this information.")
+    end
 
     # only smart proxies with V2 API are supported
```
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb]: Does this comment belong to `unless subnet.dhcp.has_capability?(:DHCP, :dhcp_filename_hostname)`?

File: app/models/concerns/orchestration/dhcp.rb [hunk 3]
```diff
@@ -82,25 +82,15 @@ def boot_server
...
-        Foreman::Deprecation.deprecation_warning('1.25', "DHCP proxy does not report dhcp_filename_* capability and reverse DNS search in Foreman will be removed.")
-        ip = NicIpResolver.new(:nic => self).to_ip_address(bs) if bs.present?
-        return ip unless ip.nil?
-        failure _("Unable to determine the host's boot server. The DHCP smart proxy failed to provide this information and this subnet is not provided with TFTP services.")
-      rescue => e
-        failure _("failed to detect boot server: %s") % e, e
-      end
-    end
+    # we no longer convert the boot server to IPv4 - smart proxy modules are required to do so if they don't support hostname
+    # first try to ask our IPv4 TFTP server for its boot server
```
Comment (ekohl) [inline app/models/concerns/orchestration/dhcp.rb]: Does this comment help? I'm not sure it really clarifies things for anyone reading this to know that in the past we did a (IP ->) DNS -> IP conversion. Today it's just passing an entry to the Smart Proxy and the Smart Proxy is responsible for any conversion if needed. That's why we have the capability check.

**Review by ekohl — COMMENTED** [review]
Code wise :+1: but I'd rephrase the commit message to explain what the patch actually does.

**Comment by ekohl** [comment]
I don't think this is really delegating it. We did that previously already but had a fallback. Also know that you don't have to fit it into a single line. https://chris.beams.io/posts/git-commit/#separate describes that you can actually write a body and I always encourage that when it makes sense.

So in this case I'd write something like:
```
Fixes #31790 - always delegate bootserver IP resolution to smart proxy

In 60c03b6a2e92c432ee66f2b27da634954c1b646c (reverse) DNS search for TFTP
server (bootServer) was delegated to the Smart Proxy if was supported, as
signaled via the dhcp_filename_hostname capability. At the same time, in
Foreman 2.0, the fallback support was deprecated. This drops the compatibility
code.
```

Context for code changes is important and really helps others understand them. It can also help yourself in the future when you do look back.

**Comment by lzap** [comment]
Relevant errors, thanks @upadhyeammit for looking into this :-)

**Comment by lzap** [comment]
Failures.

@ekohl I am just wondering, we have made the change for ISC DHCP ourselves, but I guess devs of other DHCP servers haven't noticed this. I am wondering if we should provide some sane default fallback mechanism for DHCP implementations which don't report none of the two capabilities.

File: app/models/concerns/orchestration/dhcp.rb [hunk 4]
```diff
@@ -79,28 +79,12 @@ def del_dhcp_conflicts
...
-        Foreman::Deprecation.deprecation_warning('1.25', "DHCP proxy does not report dhcp_filename_* capability and reverse DNS search in Foreman will be removed.")
-        ip = NicIpResolver.new(:nic => self).to_ip_address(bs) if bs.present?
-        return ip unless ip.nil?
-        failure _("Unable to determine the host's boot server. The DHCP smart proxy failed to provide this information and this subnet is not provided with TFTP services.")
-      rescue => e
-        failure _("failed to detect boot server: %s") % e, e
-      end
+    unless subnet.dhcp.has_capability?(:DHCP, :dhcp_filename_hostname)
+      failure _("Unable to determine boot server. The DHCP smart proxy doesn't provide dhcp_filename capability or failed to provide this information.")
     end
```
Comment (lzap) [inline app/models/concerns/orchestration/dhcp.rb:86]: I don't think this is correct, this is too strict.

I believe we should just file a warning into log if none of the `dhcp_filename_ipv4` and `dhcp_filename_hostname` are reported and that's it. We can't do anything about this and foreman should attempt to perform the orchestration. If it fails it's up to the DHCP owner to fix the component.

We can probably be verbose in the warning: _DHCP proxy does not report dhcp_filename_ipv4 or dhcp_filename_hostname capability. Foreman will pass boot server XXX as-is, the request might fail._

File: app/models/concerns/orchestration/dhcp.rb [hunk 5]
```diff
@@ -79,28 +79,12 @@ def del_dhcp_conflicts
...
-        Foreman::Deprecation.deprecation_warning('1.25', "DHCP proxy does not report dhcp_filename_* capability and reverse DNS search in Foreman will be removed.")
-        ip = NicIpResolver.new(:nic => self).to_ip_address(bs) if bs.present?
-        return ip unless ip.nil?
-        failure _("Unable to determine the host's boot server. The DHCP smart proxy failed to provide this information and this subnet is not provided with TFTP services.")
-      rescue => e
-        failure _("failed to detect boot server: %s") % e, e
-      end
+    unless subnet.dhcp.has_capability?(:DHCP, :dhcp_filename_hostname)
+      logger.warn 'DHCP proxy does not report dhcp_filename_ipv4 or dhcp_filename_hostname capability. Foreman will pass boot server XXX as-is, the request might fail.'
     end
```
Comment (lzap) [inline app/models/concerns/orchestration/dhcp.rb:86]: You have `XXX` in the string, store the bootserver (or TFTP Smart Proxy URL) in a variable and use it, then return it.

While you will be editing, can you add one more warning? When tftp.bootServer is nil, also warn:

"Using TFTP Smart Proxy hostname as the boot server name: #{boot_server}"

Thanks!

**Review by lzap — CHANGES_REQUESTED** [review]
Other than that I am good.

**Comment by lzap** [comment]
You've suffered enough, the change you are looking for is:

```ruby
diff --git a/test/models/orchestration/dhcp_test.rb b/test/models/orchestration/dhcp_test.rb
index 6ba290965..e85ff864a 100644
--- a/test/models/orchestration/dhcp_test.rb
+++ b/test/models/orchestration/dhcp_test.rb
@@ -87,7 +87,7 @@ class DhcpOrchestrationTest < ActiveSupport::TestCase
   end

   test "provision interface DHCP records should contain default filename/next-server attributes for IPv4 tftp proxy" do
-    ProxyAPI::TFTP.any_instance.expects(:bootServer).returns('192.168.1.1')
+    ProxyAPI::TFTP.any_instance.expects(:bootServer).returns('192.168.1.1').twice
     subnet = FactoryBot.build(:subnet_ipv4, :dhcp, :tftp)
     h = as_admin do
       FactoryBot.create(:host, :with_dhcp_orchestration, :with_tftp_dual_stack_orchestration, :subnet => subnet)
@@ -98,7 +98,7 @@ class DhcpOrchestrationTest < ActiveSupport::TestCase
   end

   test "provision interface DHCP records should contain PXELinux BIOS filename/next-server attributes for IPv4 tftp proxy" do
-    ProxyAPI::TFTP.any_instance.expects(:bootServer).returns('192.168.1.1')
+    ProxyAPI::TFTP.any_instance.expects(:bootServer).returns('192.168.1.1').twice
     subnet = FactoryBot.build(:subnet_ipv4, :dhcp, :tftp)
     h = as_admin do
       FactoryBot.create(:host, :with_dhcp_orchestration, :with_tftp_dual_stack_orchestration, :subnet => subnet, :pxe_loader => 'PXELinux BIOS')
```

With that we are good, thanks for patience.

**Comment by upadhyeammit** [comment]
@lzap request to review and maybe merge if we are good on this! thank you!

**Review by lzap — APPROVED** [review]
Thank you for the patience.
